src/testdrive/src/action/kafka/ingest.rs

```diff
@@ -26,33 +26,77 @@ pub struct IngestAction {
     topic_prefix: String,
     partition: i32,
     format: Format,
+    key_format: Option<Format>,
     timestamp: Option<i64>,
     publish: bool,
     rows: Vec<String>,
 }
 
 enum Format {
+    Avro { schema: String },
+    Proto { message: String },
+    Bytes,
+}
+
+enum Encoder {
     Avro {
-        key_schema: Option<String>,
-        value_schema: String,
+        schema: Schema,
+        schema_id: i32,
     },
     Proto {
-        message: String,
+        parser: &'static dyn Fn(&str) -> Result<protobuf::DynMessage, failure::Error>,
+        validator: &'static dyn Fn(&[u8]) -> Result<Box<dyn fmt::Debug>, failure::Error>,
     },
     Bytes,
 }
 
+impl Encoder {
+    fn decode_into_buf(&self, row: &str, buf: &mut Vec<u8>) -> Result<(), String> {
```

Comment:
Weird for an encoder to have a  `decode_into_buf` method. Am I crazy? Really seems like this method is encoding `row` into `buf`.